File: buildSrc/src/main/java/wolkenschloss/Destroy.java
package wolkenschloss;

import org.gradle.api.DefaultTask;
import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.file.FileSystemOperations;
import org.gradle.api.file.RegularFileProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.provider.ProviderFactory;
import org.gradle.api.tasks.Destroys;
import org.gradle.api.tasks.Internal;
import org.gradle.api.tasks.TaskAction;
import org.libvirt.LibvirtException;
import wolkenschloss.domain.DomainOperations;
import wolkenschloss.pool.PoolOperations;

import javax.inject.Inject;
import java.util.UUID;

public abstract class Destroy extends DefaultTask {

    @Destroys
    abstract public RegularFileProperty getPoolRunFile();

    @Destroys
    abstract public DirectoryProperty getBuildDir();

    @Internal
    abstract public Property<String> getDomain();

    @Inject
    abstract public FileSystemOperations getFileSystemOperations();

    @Internal
    abstract public Property<PoolOperations> getPoolOperations();

    @Internal
    abstract public Property<DomainOperations> getDomainOperations();

    @Inject
    abstract public ProviderFactory getProviderFactory();

    @TaskAction
    public void destroy() throws LibvirtException {
        destroyDomain();
        destroyPool();
        deleteBuildDirectory();
    }

    private void destroyDomain() throws LibvirtException {
        var domainOperations = getDomainOperations().get();
        var deleted = domainOperations.deleteDomainIfExists(getDomain());
        if (deleted) {
            getLogger().info("Domain deleted.");
        }
    }

    private void destroyPool() throws LibvirtException {
        if (getPoolRunFile().get().getAsFile().exists()) {
            var content = getProviderFactory().fileContents(getPoolRunFile());

            @SuppressWarnings("NullableProblems")
            var uuid = content.getAsText().map(UUID::fromString).get();

            getPoolOperations().get().destroy(uuid);
            getFileSystemOperations().delete(f -> f.delete(getPoolRunFile()));
        }
    }

    private void deleteBuildDirectory() {
        getFileSystemOperations().delete(
                spec -> spec.delete(getBuildDir()));
    }
}

File: buildSrc/src/main/kotlin/gradle/testbed/Destroy.kt
package family.haschka.wolkenschloss.gradle.testbed

import org.gradle.api.DefaultTask
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.FileSystemOperations
import org.gradle.api.provider.Property
import org.gradle.api.provider.ProviderFactory
import org.gradle.api.tasks.Destroys
import org.gradle.api.tasks.Internal
import org.gradle.api.tasks.TaskAction
import javax.inject.Inject

abstract class Destroy : DefaultTask() {

    @get:Destroys
    abstract val buildDir: DirectoryProperty

    @get:Internal
    abstract val domain: Property<String>

    @get:Inject
    abstract val fileSystemOperations: FileSystemOperations

    @get:Inject
    abstract val providerFactory: ProviderFactory

    @TaskAction
    fun destroy() {

        project.exec {
            commandLine("multipass", "exec", domain.get(), "--", "/bin/bash", "-c",
            "microk8s stop || true")
            isIgnoreExitValue = true
        }
        project.exec {
            commandLine("multipass", "delete", domain.get())
            isIgnoreExitValue = true
        }

        project.exec {
            commandLine("multipass", "purge")
        }

        fileSystemOperations.delete { delete(buildDir) }
    }
}
